modules/gapi/src/streaming/onevpl/default.cpp

```diff
@@ -0,0 +1,181 @@
+// This file is part of OpenCV project.
+// It is subject to the license terms in the LICENSE file found in the top-level directory
+// of this distribution and at http://opencv.org/license.html.
+//
+// Copyright (C) 2022 Intel Corporation
+
+#include <algorithm>
+#include <fstream>
+#include <iostream>
+#include <cctype>
+#include <tuple>
+
+#include <opencv2/gapi.hpp>
+#include <opencv2/gapi/streaming/onevpl/default.hpp>
+#include <opencv2/gapi/streaming/onevpl/source.hpp>
+
+#ifdef HAVE_ONEVPL
+
+// FIXME: find a way to remove this
+#ifdef HAVE_DIRECTX
+#ifdef HAVE_D3D11
+#pragma comment(lib,"d3d11.lib")
+
+// get rid of generate macro max/min/etc from DX side
+#define D3D11_NO_HELPERS
+#define NOMINMAX
+#include <d3d11.h>
+#pragma comment(lib, "dxgi")
+#undef NOMINMAX
+#undef D3D11_NO_HELPERS
+
+#endif // HAVE_D3D11
+#endif // HAVE_DIRECTX
+
+namespace {
+namespace cfg {
+// FIXME: Move OneVPL arguments parser to a single place
+typename cv::gapi::wip::onevpl::CfgParam create_from_string(const std::string &line) {
+    using namespace cv::gapi::wip;
+
+    if (line.empty()) {
+        throw std::runtime_error("Cannot parse CfgParam from emply line");
+    }
+
+    std::string::size_type name_endline_pos = line.find(':');
+    if (name_endline_pos == std::string::npos) {
+        throw std::runtime_error("Cannot parse CfgParam from: " + line +
+                                 "\nExpected separator \":\"");
+    }
+
+    std::string name = line.substr(0, name_endline_pos);
+    std::string value = line.substr(name_endline_pos + 1);
+
+    return cv::gapi::wip::onevpl::CfgParam::create(name, value,
+                                                   /* vpp params strongly optional */
+                                                   name.find("vpp.") == std::string::npos);
+}
+} // namespace cfg
+} // anonymous namespace
+
+namespace {
+template <typename COMNonManageableType>
+void release(COMNonManageableType *ptr) {
+    if (ptr) {
+        ptr->Release();
+    }
+}
+template <typename COMNonManageableType>
+using ComPtrGuard = std::unique_ptr<COMNonManageableType, decltype(&release<COMNonManageableType>)>;
+
+template <typename COMNonManageableType>
+ComPtrGuard<COMNonManageableType> createCOMPtrGuard(COMNonManageableType *ptr = nullptr) {
```

Comment:
DO not put this into library regular sources please.
There was Big comment in sample ( as far i remember) which warns about why this pointer type should be used in sample only: It's not actual COM pointer and should not be mixed with it and you must not stick to its correctness

Reply:
Any suggestion on extracting all the VPL info to access it for remote context, but not keeping the code in sample?

Reply:
I would propose the following solution: we have all necessary entities already to implement any behavior here

```
// use device selector to use its device selection direct purpose
shared_ptr<CfgParamDeviceSelector> default_accel_contex (new CfgParamDeviceSelector("MFX_ACCELERATION_TYPE_DX11"));

// creat source using device selector
auto source = make_vpl_source(input_file, default_accel_contex);

// get pointers for open vino remoe context using convenient device selector  methods
void*device_ptr = default_accel_contex.get_devices().begin()->get_ptr();
void* ctx = default_accel_contex.get_context()->get_ptr();
```

In this context `CfgParamDeviceSelector` represents you desired "default" behavior. All entities was created already and we must re-use them without new redundant classes & functions


**P.S.** Perhaps we would need to implement factory for device_selector creation only in the similar way as it done in `make_vpl_source` using variadic templates as selector construction arguments